File: server/public/tools/elements/views/GtSettings.ts
import {PolymerElement} from "../../polymer/PolymerElement";
import {View} from "../../client/router/View";
import {Element} from "../../polymer/Annotations";
import {GtLabel} from "../widgets/GtLabel";
import {GtCheckbox} from "../widgets/GtCheckbox";
import {GtDialog} from "../widgets/GtDialog";
import {GtButton} from "../widgets/GtButton";
import {GtBox} from "../widgets/GtBox";

type SettingsTab = [string, string, string];

@View("settings", () => [{title: "Settings", link: "/settings", active: true}])
@Element({
	selector: "gt-settings",
	template: "/assets/views/settings.html",
	dependencies: [GtBox, GtButton, GtDialog, GtCheckbox, GtLabel]
})
export class GtSettings extends PolymerElement {
	private tabs: SettingsTab[] = [
		//["General", "settings", "general"],
		//["Notifications", "notifications", "notifications"],
		//["Game", "folder", "game"],
		["Advanced", "build", "advanced"]
	];

	private tab = "advanced";

	private SelectTab(ev: PolymerModelEvent<SettingsTab>) {
		let tab = ev.model.item;
		this.tab = tab[2];
	}

	// Advanced
	private faster_loading = localStorage.getItem("loading.fast") === "1";
	private verbose_gtp3 = localStorage.getItem("socket.verbose") === "1";
	private sourcemap_fix = localStorage.getItem("traceur.sourcemap.fix") === "1";
	private streams_zerolatency = localStorage.getItem("streams.zerolatency") === "1";

	private ToggleLocalStorage(ev: Event) {
		let target = <GtCheckbox & HTMLElement> ev.target;
		let key = target.getAttribute("key");
		localStorage.setItem(key, target.checked ? "1" : "0");
	}
}

File: client/src/gt/component/misc/GtSettings.scala
package gt.component.misc

import gt.component.widget.GtBox
import gt.component.widget.form.{GtButton, GtCheckbox, GtLabel}
import gt.component.{GtHandler, Tab, View}
import util.Settings
import util.Settings.Setting
import util.annotation.data
import util.jsannotation.js
import xuen.Component
import xuen.rx.Var

object GtSettings extends Component[GtSettings](
	selector = "gt-settings",
	templateUrl = "/assets/imports/views/settings.html",
	dependencies = Seq(GtBox, GtButton, GtLabel, GtCheckbox)
) with View {
	val module = "settings"
	val tabs: TabGenerator = (selector, path, user) => Seq(
		Tab("Settings", "/settings", active = true)
	)
}

@js class GtSettings extends GtHandler {
	@data case class SettingsTab(title: String, icon: String, id: String)

	val tabs = Seq(
		//SettingsTab("General", "settings", "general"),
		//SettingsTab("Notifications", "notifications", "notifications"),
		//SettingsTab("Game", "folder", "game"),
		SettingsTab("Advanced", "build", "advanced")
	)

	val tab = Var[String]("advanced")

	val loadingFast = binding(Settings.`loading.fast`)
	val socketVerbose = binding(Settings.`socket.verbose`)

	def binding(setting: Setting[Boolean]): Var[Boolean] = {
		val rx = Var[Boolean](setting.value)
		rx ~> { state => setting := state }
		rx
	}
}
